src/charm.py

```diff
@@ -75,26 +108,109 @@ def _on_postgresql_pebble_ready(self, event: WorkloadEvent) -> None:
         else:
             self.unit.status = WaitingStatus("waiting for Pebble in workload container")
 
+    def _on_upgrade_charm(self, _) -> None:
+        # Add labels required for replication when the pod loses them (like when it's deleted).
+        self._patch_pod_labels()
+
+    def _patch_pod_labels(self) -> None:
+        """Add labels required for replication to the current pod."""
+        try:
+            client = Client()
+            patch = {
+                "metadata": {"labels": {"application": "patroni", "cluster-name": self._namespace}}
+            }
+            client.patch(
+                Pod,
+                name=self._unit_name_to_pod_name(self._unit),
+                namespace=self._namespace,
+                obj=patch,
+            )
+        except ApiError as e:
+            logger.error("failed to patch pod")
+            self.unit.status = BlockedStatus(f"failed to patch pod with error {e}")
+
+    def _create_resources(self):
+        """Create kubernetes resources needed for Patroni."""
+        try:
+            client = Client()
+            with open("src/resources.yaml") as f:
+                for obj in codecs.load_all_yaml(f, context=self._context):
+                    client.create(obj)
+        except ApiError as e:
+            logger.error("failed to create resources")
+            self.unit.status = BlockedStatus(f"failed to create resources with error {e}")
+
     def _on_get_postgres_password(self, event: ActionEvent) -> None:
         """Returns the password for the postgres user as an action response."""
         event.set_results({"postgres-password": self._get_postgres_password()})
 
+    def _on_get_primary(self, event: ActionEvent) -> None:
+        """Get primary instance."""
+        try:
+            primary = self._patroni.get_primary(unit_name_pattern=True)
+            event.set_results({"primary": primary})
+        except RetryError as e:
+            logger.error(f"failed to get primary with error {e}")
+
+    def _on_update_status(self, _) -> None:
```

Comment:
Might it be an idea to set the `ActiveStatus` message in here if the current unit is the primary? Something like:

```python
if self._patroni.get_primary(unit_name_pattern=True) == self.unit.name:
	self.unit.status = ActiveStatus("Primary")
```